File: src/background/expire.ts
/**
 * 自然过期（地图 cell 8 规则 8 · 阶段 4-B）：临时类认知老到一定程度 → 自然失效。
 *
 * 只有【临时类】（情绪 state、待验证假设 hypothesis）会自然过期：久没被印证 → 标 invalidAt
 *   （保留可溯源、不再被召回），呼应"此刻的情绪很快就该忘"。
 * 【稳定类】（明确偏好 preference、事实 fact 等）**永不自动失效**——哪怕很久没提（规则 8）。
 *
 * 纪律：失效 = 标 invalidAt 保留，不删（cell 6 失效而非删除）。这是 MemoWeft 自己按规则过期，
 *   不是用户主动删；用户纠正/删除走另路（M6 / 规则 10）。
 */
import { config } from '../config.ts';
import type { CognitionStore } from '../cognition/store.ts';

export interface ExpireDeps {
  cognitionStore: CognitionStore;
}

export interface ExpireResult {
  /** 本次自然过期（标 invalidAt）的认知条数。 */
  expired: number;
}

const DAY_MS = 86_400_000;

/** 把临时类里"距上次印证已超过期阈值"的认知标 invalidAt。稳定类不动。 */
export function expire(subjectId: string, deps: ExpireDeps, now: Date = new Date()): ExpireResult {
  const thresholds = config.background.expireAfterDays;
  let expired = 0;
  for (const c of deps.cognitionStore.active(subjectId)) {
    const days = thresholds[c.contentType];
    if (days == null) continue; // 不在过期名单 → 永不自动失效（明确偏好/事实）
    const ageDays = (now.getTime() - new Date(c.updatedAt).getTime()) / DAY_MS;
    if (ageDays > days) {
      deps.cognitionStore.update(c.id, { invalidAt: now.toISOString() });
      expired++;
    }
  }
  return { expired };
}

File: py/src/memoweft/expire.py
"""自然过期 —— 移植自 src/background/expire.ts。

临时类(state/hypothesis/trend)久没被印证 → 标 invalid_at(保留可溯源、不再召回);
稳定类(fact/preference 等,不在 expire_after_days 名单)永不自动失效。纯规则、无 LLM。
归档由 active() 排除(归档要保住可恢复,不给标失效)。
"""
from __future__ import annotations

from datetime import datetime

from .clock import epoch_ms, parse_iso_ms, to_iso_z
from .config import CONFIG, Config
from .store.cognition import SqliteCognitionStore
from .types import CognitionPatch

_DAY_MS = 86_400_000


def expire(subject_id: str, store: SqliteCognitionStore, now: datetime, cfg: Config = CONFIG) -> int:
    """把临时类里超过期阈值(严格 >)的认知标 invalid_at。稳定类不动。返回过期条数。对齐 expire.ts:28-43。"""
    thresholds = cfg.expire_after_days
    now_ms = epoch_ms(now)
    now_iso = to_iso_z(now)
    expired = 0
    for c in store.active(subject_id):
        days = thresholds.get(c.content_type)
        if days is None:
            continue  # 不在过期名单 → 永不自动失效(明确偏好/事实)
        age_days = (now_ms - parse_iso_ms(c.updated_at)) / _DAY_MS
        if age_days > days:
            store.update(c.id, CognitionPatch(invalid_at=now_iso))
            expired += 1
    return expired
